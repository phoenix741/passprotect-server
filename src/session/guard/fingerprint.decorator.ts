import { createParamDecorator } from '@nestjs/common';

export const FingerprintContext = createParamDecorator((data, [, , ctx]) => (ctx.req.fingerprint || {}).hash);
