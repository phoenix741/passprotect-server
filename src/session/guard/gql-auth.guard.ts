import { decode } from 'jsonwebtoken';
import { Injectable, ExecutionContext, UnauthorizedException } from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { GqlExecutionContext } from '@nestjs/graphql';
import { JwtPayload } from '../../users/models/user.models';

@Injectable()
export class GqlAuthGuard extends AuthGuard('jwt') {
  canActivate(context: ExecutionContext) {
    const ctx = GqlExecutionContext.create(context);
    const {
      fingerprint: { hash },
      headers: { authorization },
    } = ctx.getContext().req;
    const payload = decode(authorization.substr(7)) as JwtPayload;
    if (payload.issuer !== hash) {
      throw new UnauthorizedException(`The requester ${hash} is incompatible with the issuer ${payload.issuer}`);
    }
    return super.canActivate(context);
  }

  getRequest(context: ExecutionContext) {
    const ctx = GqlExecutionContext.create(context);
    return ctx.getContext().req;
  }
}
