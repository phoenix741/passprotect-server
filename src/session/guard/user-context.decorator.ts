import { createParamDecorator } from '@nestjs/common';

export const UserContext = createParamDecorator((data, [, , ctx]) => ctx.req.user);
