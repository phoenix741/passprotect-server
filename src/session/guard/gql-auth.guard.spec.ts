import { GqlAuthGuard } from './gql-auth.guard';

describe('GqlAuthGuard', () => {
  let guard;
  let context;
  let req;

  beforeEach(() => {
    guard = new GqlAuthGuard();
    req = {
      session: 'session',
      user: 'user',
      fingerprint: {
        hash: 'fingerprint1',
      },
      headers: {
        authorization: 'bearer ' +
          'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.' +
          'eyJzdWIiOiJJRElEIiwiaXNzdWVyIjoiMjRhNzM1MmM5MTlkMzUyMzNlNWQ4M2RhYWZmNDQzNmMiLCJpY' +
          'XQiOjE1NjQzMzc2MDMsImV4cCI6MTU2NDMzODUwMywianRpIjoiNWQzZGU1YzMyNWU3NjIzNzE1NGQ3MjM' +
          'xIn0.xAlbFEAh9Sk3ck47D54y6jB6Qy9d6pgof7jkn8OELFo',
      },
    };
    context = {
      getArgByIndex(idx) {
        return idx;
      },
      getContext() {
        return {
          req,
        };
      },
    };
  });

  describe('canActivate', () => {
    it('hash equals', () => {
      req.fingerprint.hash = '24a7352c919d35233e5d83daaff4436c';
      guard.canActivate(context);
    });

    it('hash different', () => {
      try {
        guard.canActivate(context);

        throw new Error('Hash should be different');
      } catch (err) {
        expect(err.message).toEqual({
          error: 'Unauthorized',
          message: 'The requester fingerprint1 is incompatible with the issuer 24a7352c919d35233e5d83daaff4436c',
          statusCode: 401,
        });
      }
    });
  });

  describe('getRequest', () => {
    it('convert express context to graphql', () => {
      expect(guard.getRequest(context)).toMatchSnapshot('request of the context');
    });
  });
});
