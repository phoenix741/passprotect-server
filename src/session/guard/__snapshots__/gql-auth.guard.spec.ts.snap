// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GqlAuthGuard getRequest convert express context to graphql: request of the context 1`] = `
Object {
  "fingerprint": Object {
    "hash": "fingerprint1",
  },
  "headers": Object {
    "authorization": "bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJJRElEIiwiaXNzdWVyIjoiMjRhNzM1MmM5MTlkMzUyMzNlNWQ4M2RhYWZmNDQzNmMiLCJpYXQiOjE1NjQzMzc2MDMsImV4cCI6MTU2NDMzODUwMywianRpIjoiNWQzZGU1YzMyNWU3NjIzNzE1NGQ3MjMxIn0.xAlbFEAh9Sk3ck47D54y6jB6Qy9d6pgof7jkn8OELFo",
  },
  "session": "session",
  "user": "user",
}
`;
